resdk/data_upload/annotate_samples.py

```diff
@@ -0,0 +1,139 @@
+"""Annotate samples in a collection."""
+
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import logging
+
+from resdk.data_upload.samplesheet import FileExporter, FileImporter
+
+__all__ = ('annotate_samples', 'export_annotation')
+
+logger = logging.getLogger(__name__)
+
+
+def annotate_samples(collection, samplesheet_source, schema='sample'):
+    """Batch-annotate the samples in a collection.
+
+    This method applies annotations to the samples in a collection based on a
+    sample annotation spreadsheet. The entries in the spreadsheet and the
+    samples in the collection are matched by sample name, and validated for
+    upload.
+
+    :param str samplesheet_source: path to a local sample annotation
+        spreadsheet, or name of one in the collection, or an existing
+        FileImporter object created from an upload
+    :param str schema: slug of the descriptor schema to use
+    """
+    # Extract the information from a sample annotation spreadsheet
+    if isinstance(samplesheet_source, FileImporter):
+        samplesheet = samplesheet_source
+    else:
+        samplesheet = FileImporter(samplesheet_source)
+
+    # TODO: Try to pull the samplesheet as a data object from the collection
+    #       once it is possible to upload samplesheets
+    # else:
+    #     sheet = collection.data.get(name=samplesheet_source,
+    #                                 type='data:annotation')
+    #     sheet.download()
+    #     samplesheet = FileImporter(sheet.files()[0])
+    #     err_samples = []
+
+    # Report summary of invalid annotations
+    err_samples = samplesheet.invalid_samples
+    if err_samples:
+        logger.error(
+            "\nInvalid annotations were provided for the following samples: %s."
+            "\nPlease fill in all bolded columns of the template "
+            "generated by the `export_annotation` method of"
+            " your collection.",
+            ', '.join(err_samples)
+        )
+
+    # Match annotation entries to collection samples
+    matched = map(
+        lambda entry: (entry, collection.samples.filter(name=entry.name)),
```

Comment:
Because of the danger of overwriting all samples in a collection in the event of a `.filter(name='')` query, I'm just going to make it so that a sample annotation will not be applied if multiple samples are queried here.